### PR TITLE
Linux ci - Compile KVM with goto-cc

### PR DESCRIPTION
**Pull request by nmanthey**
This change adds a github workflow that compiles parts of the Linux kernel with goto-cc, and fails in case compilation fails. Furthermore, the input fails that lead to the failure are archived, to allow easy consumption and investigation.

### Note: This is currently failing, and should help to improve goto-cc to be able to compile Linux

An example failed run with the broken input as artifact has been created as part of my testing: https://github.com/nmanthey/cbmc/pull/5/checks?check_run_id=2093209018

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [N/A] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [N/A] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [N/A] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [N/A] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: .github/workflows/build-and-test-Linux.yaml
Comment (tautschnig) [inline .github/workflows/build-and-test-Linux.yaml:53]: How does one actually download these? GitHub keeps telling me that I need to have the "actions" scope, perhaps the repository permissions are set up in a way that doesn't permit non-members to do this?

File: .github/workflows/build-and-test-Linux.yaml
Comment (tautschnig) [inline .github/workflows/build-and-test-Linux.yaml]: As far as I recall, we just need `build-essential libssl-dev` for the Linux kernel.

File: .github/workflows/build-and-test-Linux.yaml
Comment (tautschnig) [inline .github/workflows/build-and-test-Linux.yaml]: I'd suggest to always use the latest release (at the risk of builds surprisingly breaking) and perhaps also using tar/zip instead of cloning?

File: .github/workflows/build-and-test-Linux.yaml
Comment (tautschnig) [reply .github/workflows/build-and-test-Linux.yaml:53]: How about doing `tail -n +1 CPROVER/faultyInput/*` instead? Really simple and avoids any permissions problems.

File: .github/workflows/build-and-test-Linux.yaml
Comment (nmanthey) [reply .github/workflows/build-and-test-Linux.yaml]: IIRC, that depends on the configuration. I can minimize this set.

File: .github/workflows/build-and-test-Linux.yaml
Comment (nmanthey) [reply .github/workflows/build-and-test-Linux.yaml:53]: On the checks page (here, e.g.: https://github.com/diffblue/cbmc/pull/5916/checks there is an "artifacts" button on the top left, which allows (me) to download, I did not see any permission restrictions in the description of the action). Yes, I could also dump the content of the files, but that likely floods the output, as those are preprocessed already.

File: .github/workflows/build-and-test-Linux.yaml
Comment (nmanthey) [reply .github/workflows/build-and-test-Linux.yaml]: If people are willing to take the risk, and address the issue (or roll this action back to the last known working release) I can build logic to always access the latest release.

On the other hand, I would like to be able to ensure that a specific version is always working. 

This is the risk I see when always using the latest release: Once the latest release fails, people might ignore the failure or disable this action to unblock themselves. Afterwards, even compiling older relases might start to break again. Hence, I decided to pick a specific release for now.

Comment (tautschnig) [reply .github/workflows/build-and-test-Linux.yaml]: Maybe we should have two such actions? One with a fixed, longterm-maintenance release (5.10 seems like a reasonable candidate) and one using the "master" branch?

File: .github/workflows/build-and-test-Linux.yaml
Comment (tautschnig) [reply .github/workflows/build-and-test-Linux.yaml:53]: Err, thank you, that was too easy. A first fix is in https://github.com/diffblue/cbmc/pull/5928.

**Comment by tautschnig** [comment]
We will next need #5730 to be merged.

**Comment by tautschnig** [comment]
> We will next need #5730 to be merged.

And then #5723, at which point the job should actually succeed.

**Comment by tautschnig** [comment]
> > We will next need #5730 to be merged.
> 
> And then #5723, at which point the job should actually succeed.

Both of which are now merged, I believe this test should now succeed.

File: .github/workflows/build-and-test-Linux.yaml
Comment (tautschnig) [inline .github/workflows/build-and-test-Linux.yaml]: May I suggest the following variation of this:
```
cat > kvm-config <<EOF
CONFIG_64BIT=y
CONFIG_X86_64=y
CONFIG_HIGH_RES_TIMERS=y
CONFIG_MULTIUSER=y
CONFIG_NET=y
CONFIG_VIRTUALIZATION=y
CONFIG_HYPERVISOR_GUEST=y
CONFIG_PARAVIRT=y
CONFIG_KVM_GUEST=y
CONFIG_KVM=y
CONFIG_KVM_INTEL=y
CONFIG_KVM_AMD=y
EOF
KCONFIG_ALLCONFIG=kvm-config make allnoconfig
make bzImage -j $(nproc)
```

File: .github/workflows/build-and-test-Linux.yaml
Comment (tautschnig) [inline .github/workflows/build-and-test-Linux.yaml]: With the above, perhaps instead do:
```
rm arch/x86/kernel/kvm.o
find arch/x86/kvm/ -name "*.o" -delete
```

File: .github/workflows/build-and-test-Linux.yaml
Comment (tautschnig) [inline .github/workflows/build-and-test-Linux.yaml]: I don't think we we should be ignoring failures here, thus drop `-k || true`. Also, shouldn't one-line-scan use (at least) the same `-j` as the command being run?

File: .github/workflows/build-and-test-Linux.yaml
Comment (nmanthey) [reply .github/workflows/build-and-test-Linux.yaml]: Sure, I can re-iterate

File: .github/workflows/build-and-test-Linux.yaml
Comment (nmanthey) [reply .github/workflows/build-and-test-Linux.yaml]: Will iterate

File: .github/workflows/build-and-test-Linux.yaml
Comment (nmanthey) [reply .github/workflows/build-and-test-Linux.yaml]: Yes, that should work.

The '-j' for one-line-scan is limited to not consume too much memory. For github, that should be good enough. For other use cases, that might be higher.

File: .github/workflows/build-and-test-Linux.yaml
Comment (nmanthey) [reply .github/workflows/build-and-test-Linux.yaml]: Let's do the stable one for now, and iterate once that one is in.

**Comment by nmanthey** [comment]
I addressed the comments, and factored out the actual compilation into a separate script, so that these steps can be executed in other environments more easily.

**Comment by tautschnig** [comment]
> I addressed the comments, and factored out the actual compilation into a separate script, so that these steps can be executed in other environments more easily.

Thanks a lot, I'll review whether I already have a patch for the invariant failure.

**Comment by tautschnig** [comment]
> > I addressed the comments, and factored out the actual compilation into a separate script, so that these steps can be executed in other environments more easily.
> 
> Thanks a lot, I'll review whether I already have a patch for the invariant failure.

A proposal for the patch is in #5967, but I need to reconsider whether that's the correct approach (it is borrowed from #5738, which also made a number of other changes).

**Review by NlightNFotis — COMMENTED** [review]
This looks good to me (modulo the CI failure).

I will wait for this to be fixed then I will approve this to get in.

**Comment by tautschnig** [comment]
@nmanthey #5967 is now merged, I believe this should pass once rebased.

**Review by tautschnig — APPROVED** [review]

File: .github/workflows/build-and-test-Linux.yaml
Comment (TGWDB) [inline .github/workflows/build-and-test-Linux.yaml:6]: I'm a little hesitant about (another) CI job that runs on all PRs, specially one that we expect to report failures.

Comment (nmanthey) [reply .github/workflows/build-and-test-Linux.yaml:6]: Feel free to suggest another branch pattern. I can also target the develop and production branch only. That might result in late surprises.

Comment (tautschnig) [reply .github/workflows/build-and-test-Linux.yaml:6]: @TGWDB Why would you expect failures here?

Comment (TGWDB) [reply .github/workflows/build-and-test-Linux.yaml:6]: @nmanthey I'm no git expert and don't have a better suggestion. I agree that late failures are not a good idea.

@tautschnig  Based on the original PR text:

"This change adds a github workflow that compiles parts of the Linux kernel with goto-cc, and fails in case compilation fails. Furthermore, the input fails that lead to the failure are archived, to allow easy consumption and investigation."

I'm happy if this works without failures, but was (mis?)reading that we expect this to fail more often due to the target and perhaps related aspects of `cbmc`.

If we expect `goto-cc` to never fail this except for when we break something then I have no objection.

**Review by TGWDB — APPROVED** [review]

**Pull request merged**
